File: ios/Classes/DocumentoscopyFlutterSdkPlugin.swift
import Flutter
import UIKit
import CSDocumentoscopySDK

public class DocumentoscopyFlutterSdkPlugin: NSObject, FlutterPlugin {
    
    private let LOG_TAG = "[CSDocumentosCopyFlutter]"
    private var flutterResult: FlutterResult?;
    
    private func resetFlutterResult() -> Void {
        self.flutterResult = nil
    }
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "documentoscopy_flutter_sdk", binaryMessenger: registrar.messenger())
        let instance = DocumentoscopyFlutterSdkPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "getPlatformVersion":
            result("iOS " + UIDevice.current.systemVersion)
        case "openCSDocumentosCopy":
            if let arguments = call.arguments as? [String:Any] {
                let sdkParams: NSDictionary = [
                    "clientId": arguments["clientId"] as! String,
                    "clientSecretId": arguments["clientSecretId"] as! String,
                    "identifierId": arguments["identifierId"] as! String,
                    "cpf": arguments["cpf"] as! String,
                ];
                
                openCSDocumentosCopy(sdkParams: sdkParams, resultParam: result)
            } else {
                result(FlutterError(code: "MissingSdkParams", message: "Unable to get sdk params from payload", details: nil))
            }
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    // TODO -> Check how to properly get params.
    private func openCSDocumentosCopy(sdkParams: NSDictionary, resultParam: @escaping FlutterResult) {
        if let result = self.flutterResult {
            // Means that we are already running and somehow the button got triggered again.
            // In this case just return.
            
            return
        }
        
        if let clientId = sdkParams["clientId"] as? String, let clientSecretId = sdkParams["clientSecretId"] as? String, let identifierId = sdkParams["identifierId"] as? String, let cpf = sdkParams["cpf"] as? String {
                   
            self.flutterResult = resultParam

            DispatchQueue.main.async {
               let sdk = CSDocumentoscopy()

               if let viewController = UIApplication.shared.keyWindow?.rootViewController {
                   sdk.delegate = self
                   sdk.initialize(clientId: clientId, clientSecret: clientSecretId, identifierId: identifierId, cpf: cpf, viewController: viewController)
               } else {
                   resultParam(FlutterError(code: "ViewControllerMissing", message: "Unable to find view controller", details: nil))
               }
            }
        } else {
            resultParam(FlutterError(code: "MissingParameters", message: "Missing clientId, clientSecretId, identifierId or CPF or all of them", details: nil))
        }
    }
}

extension DocumentoscopyFlutterSdkPlugin: CSDocumentoscopyDelegate {
    public func didOpen() {
        NSLog("\(LOG_TAG) - called didOpen")
    }
    
    public func didTapClose() {
        NSLog("\(LOG_TAG) - called didTapClose")
        
        if let result = self.flutterResult {
            result(FlutterError(code: "UserCancel", message: "UserCancel", details: nil))
        }
        
        self.resetFlutterResult()
    }
    
    public func didFinishCapture(result: CSDocumentoscopyResult) {
        NSLog("\(LOG_TAG) - called didFinishCapture")
        
        let responseMap = NSMutableDictionary();
        responseMap.setValue(result.documentType.rawValue, forKey: "documentType")
        responseMap.setValue(result.sessionId, forKey: "sessionId")
        
        if let result = self.flutterResult {
            result(responseMap)
        }
        
        self.resetFlutterResult()
    }
    
    public func didReceiveError(error: CSDocumentoscopyError) {
        NSLog("\(LOG_TAG) - called didReceiveError")
        
        if let result = self.flutterResult {
            result(FlutterError(code: String(error.errorCode), message: String(error.text), details: nil))
        }
        
        self.resetFlutterResult()
    }
}

File: android/src/main/kotlin/br/com/clearsale/documentoscopy_flutter_sdk/DocumentoscopyFlutterSdkPlugin.kt
package br.com.clearsale.documentoscopy_flutter_sdk

import android.app.Activity
import android.content.Context
import android.graphics.Color
import android.util.Log
import com.clear.studio.csdocs.entries.CSDocumentoscopy
import com.clear.studio.csdocs.entries.CSDocumentoscopySDK
import com.clear.studio.csdocs.entries.CSDocumentoscopySDKColorsConfig
import com.clear.studio.csdocs.entries.CSDocumentoscopySDKConfig
import com.clear.studio.csdocs.entries.CSDocumentoscopySDKError
import com.clear.studio.csdocs.entries.CSDocumentoscopySDKListener
import com.clear.studio.csdocs.entries.CSDocumentoscopySDKResult
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/** DocumentoscopyFlutterSdkPlugin */
class DocumentoscopyFlutterSdkPlugin: FlutterPlugin, MethodCallHandler, ActivityAware {
  /// The MethodChannel that will the communication between Flutter and native Android
  ///
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
  private lateinit var channel : MethodChannel
  private var activity: Activity? = null
  private var context: Context? = null

  private var logTag = "[CSDocumentosCopy]"

  var flutterResult: Result? = null

  override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    context = flutterPluginBinding.applicationContext
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "documentoscopy_flutter_sdk")
    channel.setMethodCallHandler(this)
  }

  override fun onMethodCall(call: MethodCall, result: Result) {
    if (call.method == "openCSDocumentosCopy") {
      openCSDocumentosCopy(call, result)
    } else {
      result.notImplemented()
    }
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }

  override fun onAttachedToActivity(binding: ActivityPluginBinding) {
    activity = binding.activity
  }

  override fun onDetachedFromActivityForConfigChanges() {
    activity = null
  }

  override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
    activity = binding.activity
  }

  override fun onDetachedFromActivity() {
    activity = null
  }

  private fun resetResult() {
    flutterResult = null;
  }

  private fun openCSDocumentosCopy(call: MethodCall, result: Result) {
    if (flutterResult !== null) {
      // Means that we are already running and somehow the button got triggered again.
      // In this case just return.

      return;
    }

    try {
      flutterResult = result;

      val clientId = call.argument<String>("clientId")
      val clientSecretId = call.argument<String>("clientSecretId")
      val identifierId = call.argument<String>("identifierId")
      val cpf = call.argument<String>("cpf")
      val primaryColor = call.argument<String>("primaryColor")
      val secondaryColor = call.argument<String>("secondaryColor")
      val tertiaryColor = call.argument<String>("tertiaryColor")
      val titleColor = call.argument<String>("titleColor")
      val paragraphColor = call.argument<String>("paragraphColor")

      // Now validate
      if (clientId.isNullOrBlank()) throw Exception("clientId is required")
      if (clientSecretId.isNullOrBlank()) throw Exception("clientSecretId is required")
      if (identifierId.isNullOrBlank()) throw Exception("identifierId is required")
      if (cpf.isNullOrBlank()) throw Exception("cpf is required")


      val sdkConfig = CSDocumentoscopySDKConfig(
        colors = CSDocumentoscopySDKColorsConfig(
          primaryColor = if (!primaryColor.isNullOrBlank()) Color.parseColor(primaryColor) else null,
          secondaryColor = if (!secondaryColor.isNullOrBlank()) Color.parseColor(secondaryColor) else null,
          tertiaryColor = if (!tertiaryColor.isNullOrBlank()) Color.parseColor(tertiaryColor) else null,
          titleColor = if (!titleColor.isNullOrBlank()) Color.parseColor(titleColor) else null,
          paragraphColor = if (!paragraphColor.isNullOrBlank()) Color.parseColor(paragraphColor) else null
        )
      )
      val csDocumentosCopyConfig = CSDocumentoscopy(clientId, clientSecretId, identifierId, cpf, sdkConfig)
      val listener = object: CSDocumentoscopySDKListener {
        override fun didFinishCapture(result: CSDocumentoscopySDKResult) {
          Log.d(logTag, "Called didFinishCapture");

          val responseMap = HashMap<String, String?>()
          responseMap["documentType"] = result.documentType?.toString()
          responseMap["sessionId"] = result.sessionId

          Log.d(logTag, "Result is $responseMap")

          flutterResult!!.success(responseMap)
          resetResult()
        }

        override fun didOpen() {
          // Nothing to do here.
          Log.d(logTag, "Called didOpen");
        }

        override fun didReceiveError(error: CSDocumentoscopySDKError) {
          Log.e(logTag, "Called didReceiveError", null);

          flutterResult!!.error(error.errorCode.toString(), error.text, null)
          resetResult()
        }

        override fun didTapClose() {
          Log.d(logTag, "Called didTapClose");

          result.error("UserCancel", "UserCancel", null)
          resetResult()
        }
      }

      if (activity?.application != null) {
        CSDocumentoscopySDK.initialize(activity!!.application, csDocumentosCopyConfig, listener)
      } else {
        throw Exception("Missing application from current activity")
      }
    } catch (t: Throwable) {
      Log.e(logTag, "Error starting CSDocumentosCopySDK", t)
      result.error("SDKError", "Failed to start CSDocumentosCopySDK", t)

      resetResult()
    }
  }
}
